### ISSUE TITLE
Print backtrace of all threads

### ISSUE THREAD

**Issue by gavv** [roc-streaming/roc-toolkit#565]
die_gracefully() is invoked from signal handler (crash_handler.cpp) or panic (panic.cpp).

See https://github.com/roc-streaming/roc-toolkit/blob/develop/src/internal_modules/roc_core/target_posix/roc_core/die.cpp

Currently it prints backtrace of current thread and aborts program.

We need to update it to print backtrace of all threads, not just current one.

It can be achieved by sending some signal (e.g. SIGQUIT) to all threads, which will cause each thread to print its backtrace and terminate. And then exiting after all threads are terminated.

Backtrace of original failed thread should be marked somehow to indicate that it was the source of the problem.

**Comment by jeshwanthreddy13** [comment]
Hi @gavv, I would like to work on this . Can I get it assigned?

**Comment by gavv** [comment]
Sure, thanks!

**Comment by gavv** [comment]
BTW do you have an idea of implementation already? I guess the tricky part is who and how decides that it's time to exit process.

**Comment by jeshwanthreddy13** [comment]
> BTW do you have an idea of implementation already? 

Initially the implementation I had in mind was same as what you had described in this issues discription i.e. the thread which receives the signal would send a SIGQUIT  to all other threads of the same process and terminate after the other threads print their call stack and terminate.

The problem with this is, there doesn't seem to be a reliable way to send signals to other threads via the program itself,  also no methods in the pthread library offer such functionality. 
Not sure if it could be implemented using libunwind. 
Another method is that we iterate over the `/proc/<pid>/task` directory and send a `SIGQUIT` signal to every thread id.

Not sure If I'm headed in the right direction. Please correct me if I have missed any other implementation. 

>I guess the tricky part is who and how decides that it's time to exit process.

When to exit could be taken care by sending SIGQUIT to all other threads and wait for their backtrace to be printed by having a shared variable that increments when a thread prints it backtrace. Also edge cases need to be taken care.

**Comment by gavv** [comment]
You can send signal to a thread by TID using tgkill() call. As for listing threads, I also don't know any other way except /proc/self. Both are linux-specific, but this is fine.

Pthread has pthread_kill() but we can't use it as we don't create all threads by ourselves and thus don't have pthread_t for all of them.

So we can read /proc and use tgkill to send SIGQUIT to each thread.

We are doing this in signal handler, which complicates things. We can't use stdio, for example.

It's also not clear how to implement the waiting reliably. What if we sent a signal to a thread, but it terminated and didn't notify us? Or what if new thread was created while we were waiting? I guess we'll need some timeout and retries.

I'm not sure if it would work, but probably a simpler approach is possible. Instead of waiting, thread can print its backtrace, then find one another running thread, send SIGQUIT to it, and exit from thread. Another thread gets SIGQUIT, and the procedure repeats until the last thread exits. However this needs experimenting I guess.

**Comment by jeshwanthreddy13** [comment]
This approach seems simple and better. I'll test it out with some edge cases and make sure it works well.

**Comment by jeshwanthreddy13** [comment]
Hi @gavv, I came across `killpg` which I assume would suit us well. 
It would send `SIGQUIT` to a process group, so to all the threads in the process also, with this we can avoid the overhead of each thread iterating through `/proc/<self>/task` to get the TID of an other thread that is alive and then kill it.

Please let me know if you think there are any shortcomings with this implementation.

**Comment by gavv** [comment]
Shells with job control (sh, bash, etc) create one process group for the whole pipeline (like "foo | bar"), so killpg would also kill other processes in the pipeline, which would be odd.

We could create our own process group, but I think this may break some shell features because it may send some signals to the process group. 

Also I wonder if killpg() delivers signal to every thread? I'd expect that it delivers it to only one thread in each process.

**Comment by jeshwanthreddy13** [comment]
> Shells with job control (sh, bash, etc) create one process group for the whole pipeline (like "foo | bar"), so killpg would also kill other processes in the pipeline, which would be odd.
  
> We could create our own process group, but I think this may break some shell features because it may send some signals to the process group.

 Makes sense, did not consider this situation. This implantation would break few shell features. 

> Also I wonder if killpg() delivers signal to every thread? I'd expect that it delivers it to only one thread in each process.

Yea, I'm not completely sure. I assume it sends signals to all threads, or even if it sends a signal to one thread we make sure every thread calls it so that all threads are terminated at the end.

**Comment by gavv** [comment]
The second point is good idea, but yes, I think we should avoid killpg in order not to break shells.

**Comment by jeshwanthreddy13** [comment]
Hi @gavv, while testing some changes I can across few issues:

1) One thing I noticed is that there are a few threads related to other modules (such as PulseAudio and liburing) that do not seem to be handled by via our signal handlers. When they are sent `SIGQUIT` they cause the main thread/process to also SIGQUIT and hence the whole process becomes defunct(since main thread also exits via pthread_exit()) and later the backtrace that gets printed is garbage.

```
$ ps -T -p 107424
    PID    SPID TTY          TIME CMD
 107424  107424 pts/0    00:00:01 roc-send
 107424  107425 pts/0    00:00:00 iou-sqp-107424
 107424  107426 pts/0    00:00:00 roc-send
 107424  107427 pts/0    00:00:00 roc-send
 107424  107428 pts/0    00:00:00 threaded-ml
```

We can probably mitigate this issue by making the main thread print its backtrace, then make it ignore `SIGQUIT` and make it sleep. Then as usual we send 'SIGQUIT' to other threads apart the main thread. At the end it receives a `SIGABRT` from the last thread exiting the signal handler and terminates.

2) As mentioned above, few threads don't go through our signal handlers when they are sent `SIGQUIT` and hence there is always an uncertainty that all our threads are not sent `SIGQUIT`(chain might be broken). 

To mitigate this issue we make sure that after a thread, lets say 'Thread A', sends SIGQUIT to another thread, lets say 'Thread B', we wait for 'Thread B' to toggle a global flag inside of 'die_gracefully()' which lets 'Thread A' know that 'Thread B' has entered the signal handler, hence can exit. If 'Thread B' is not handled via our signal handler we call the next Thread available. 

Please let me know if I'm missing anything or wrong somewhere.

**Comment by gavv** [comment]
Thanks for the analysis. This is getting complicated!

The idea of having some thread waiting for other threads bothers me. We're inside signal handler and very limited in what we can do safely, and also a thread that we're waiting for may exit or crash or block signals (BTW this problem is relevant for all approaches discussed so far).

So, based on all the above discussion, maybe take a more blunt approach?

Instead of playing tricks with signals, we could implement an approach similar to what most debuggers do on crash:

1. SIGQUIT is sent to the process; random thread handles it
2. the thread calls `vfork()`
3. as the effect, the parent thread is blocked (other threads are not blocked); also, the parent process stops handling signals;
4. in child process, we immediately perform `ptrace(PTRACE_ATTACH)` on the parent process
5. now, all threads of the parent process are paused
6. in child process, we use [libunwind-ptrace](https://www.nongnu.org/libunwind/man/libunwind-ptrace(3).html) + [libunwind remote unwinding](https://www.nongnu.org/libunwind/man/libunwind(3).html) to obtain and print backtrace of every thread of parent process (probably we need to call ptrace for every tid? not sure)
7. then child process kills parent and itself

It would require to write some code using libunwind, but most of the code can be copied from existing `print_backtrace_safe()` function (we just need to replace local unwinding with remote). But this is probably the only reliable way to implement it.

Here we rely on linux-specific behavior of vfork (see "Linux description" [here](https://man7.org/linux/man-pages/man2/vfork.2.html)), so even better we can use `clone()` with `VFORK` flag, which is same as vfork, but makes it explicit that this code is linux-only.

(We could actually use regular fork instead of vfork, but it would be more racy, as the parent process can die from another signal while we're forking)

Thoughts?

**Comment by jeshwanthreddy13** [comment]
> Thanks for the analysis. This is getting complicated!
> 
> The idea of having some thread waiting for other threads bothers me. We're inside signal handler and very limited in what we can do safely, and also a thread that we're waiting for may exit or crash or block signals (BTW this problem is relevant for all approaches discussed so far).
> 
> So, based on all the above discussion, maybe take a more blunt approach?
> 
> Instead of playing tricks with signals, we could implement an approach similar to what most debuggers do on crash:
> 
> 1. SIGQUIT is sent to the process; random thread handles it
> 2. the thread calls `vfork()`
> 3. as the effect, the parent thread is blocked (other threads are not blocked); also, the parent process stops handling signals;
> 4. in child process, we immediately perform `ptrace(PTRACE_ATTACH)` on the parent process
> 5. now, all threads of the parent process are paused
> 6. in child process, we use [libunwind-ptrace](https://www.nongnu.org/libunwind/man/libunwind-ptrace(3).html) + [libunwind remote unwinding](https://www.nongnu.org/libunwind/man/libunwind(3).html) to obtain and print backtrace of every thread of parent process (probably we need to call ptrace for every tid? not sure)
> 7. then child process kills parent and itself
> 

> It would require to write some code using libunwind, but most of the code can be copied from existing `print_backtrace_safe()` function (we just need to replace local unwinding with remote). But this is probably the only reliable way to implement it.

Yes, using remote unwind can help us achieve this in a very reliable way.

> Here we rely on linux-specific behavior of vfork (see "Linux description" [here](https://man7.org/linux/man-pages/man2/vfork.2.html)), so even better we can use `clone()` with `VFORK` flag, which is same as vfork, but makes it explicit that this code is linux-only.
> 
> (We could actually use regular fork instead of vfork, but it would be more racy, as the parent process can die from another signal while we're forking)
> 
> Thoughts?

This approach seems to be better and more reliable since there is no need of waiting in the signal handlers and we will be having control over all of the threads. 
Thanks, I will start implementing this.
